src/app/clusters/service-area-server/service-area-cluster-objects.h

```diff
@@ -0,0 +1,368 @@
+    }
+
+    /**
+     * @brief This is a full constructor that initializes the location object with the given values. All values are deep copied.
+     * @param[in] aLocationId The unique identifier of this location.
+     * @param[in] aMapId The identifier of the supported map associated with this location.
+     * @param[in] aLocationName A human readable name for this location (empty string if not used).
+     * @param[in] aFloorNumber The floor level of this location - use negative values for below ground.
+     * @param[in] aAreaTypeTag A common namespace Area tag - indicates an association of the location with an indoor or outdoor area
+     * of a home.
+     * @param[in] aLandmarkTag A common namespace Landmark tag - indicates an association of the location with a home landmark.
+     * @param[in] aPositionTag A common namespace Position tag - indicates the position of the location with respect to the
+     * landmark.
+     * @param[in] aSurfaceTag A common namespace Floor Surface tag - indicates an association of the location with a surface type.
+     *
+     * @note Requirements regarding what combinations of fields and values are valid are not checked by this class.
+     * @note If aLocationName is larger than kLocationNameMaxSize, it will be truncated.
+     * @note If aLocationName is an empty string and aFloorNumber and aAreaTypeTag are null, locationInfo will be set to null.
+     */
+    LocationStructureWrapper(uint32_t aLocationId, const DataModel::Nullable<uint8_t> & aMapId, const CharSpan & aLocationName,
+                             const DataModel::Nullable<int16_t> & aFloorNumber,
+                             const DataModel::Nullable<AreaTypeTag> & aAreaTypeTag,
+                             const DataModel::Nullable<LandmarkTag> & aLandmarkTag,
+                             const DataModel::Nullable<PositionTag> & aPositionTag,
+                             const DataModel::Nullable<FloorSurfaceTag> & aSurfaceTag)
+    {
+        Set(aLocationId, aMapId, aLocationName, aFloorNumber, aAreaTypeTag, aLandmarkTag, aPositionTag, aSurfaceTag);
+    }
+
+    /**
+     * @brief This is a copy constructor that initializes the location object with the values from another location object. All
+     * values are deep copied.
+     * @param[in] aOther The location object to copy.
+     *
+     * @note If the locationName is empty string and aFloorNumber and aAreaTypeTag are null, locationInfo will be set to null.
+     */
+    LocationStructureWrapper(const LocationStructureWrapper & aOther) { *this = aOther; }
+
+    /**
+     * @brief This is an assignment operator that initializes the location object with the values from another location object. All
+     * values are deep copied.
+     * @param[in] aOther The location object to copy.
+     *
+     * @note If the locationName is empty string and aFloorNumber and aAreaTypeTag are null, locationInfo will be set to null.
```

Comment:
why is this assignment (and subsequently copied over to the copy constructor) here?

I would expect a copy constructor to copy the object as-is not transform it.

Reply:
Should the copy constructor make an exact copy of an object that violates requirements?

The alternative would be to throw an exception?

Reply:
If the original object has a non-null locationInfo struct but all of it's values are null, the copy results in an object that has the same information but the values are slightly different.

Note that in this case the original object is not a valid object according to requirements.  The alternative would be to assert or throw an exception because the original object is illegal.

Although the edge case object is illegal, because of the redundancy between a null structure and an instantiated structure that has only null or empty string contents it's clear what the meaning of the illegal structure should be. This case is unlikely to ever occur, but the code has to handle all possibilities and it seems better to modify the syntax of the object slightly to make it legal as opposed to propagating and illegal object or 'crashing' the program (likely result of assert or exception).

Reply:
I will re-state this: I fully expect `operator=` to create an identical copy.

if you do not create an identical copy (it seems you do not) give it a method name and do not use `operator=`.

Reply:
Moved to #34519